programs/vote/src/vote_state.rs

```diff
@@ -373,6 +374,70 @@ impl VoteState {
         self.last_timestamp = BlockTimestamp { slot, timestamp };
         Ok(())
     }
+
+    /// Find slots that would be slashable between the two vote states
+    /// This function is used outside of the bank to check if any concurrent
+    /// forks have slashable votes.
+    pub fn slashable_slots(&self, other: &VoteState) -> Vec<Slot> {
+        let mut slashable = vec![];
+        slashable.extend(self.votes());
+        slashable.extend(other.votes());
+        let mut test = VoteState::default();
+        slashable_votes.sort();
+        slashable_votes.dedup();
+        for s in slashable_votes {
+            test.process_vote_unchecked(s);
+        }
+        let mut retval = vec![];
+        if test.votes != self.votes && other.older(self) {
+            let test_set: HashSet<_> = test.votes.iter().cloned().collect();
+            let self_set: HashSet<_> = self.votes.iter().cloned().collect();
+            retval.extend(test_set.difference(self_set));
+        }
+        if test.votes != other.votes && self.older(other) {
+            let test_set: HashSet<_> = test.votes.iter().cloned().collect();
+            let b_set: HashSet<_> = other.votes.iter().cloned().collect();
+            retval.extend(test_set.difference(b_set));
+        }
+        retval
+    }
+
+    /// Check which VoteState has the oldest newest vote
+    fn older(&self, other: &VoteState) -> bool {
+        self.votes.last().unwrap_or(self.root) < other.votes.last().unwrap_or(other.root);
+    }
+
+    /// Set the VoteState as slashed if the slot is slashable
+    /// * slashable_slot - must be a slot which is not an ancestor of this bank
+    /// Instruction must validate that the slashable_slot is not an ancestor of this bank,
+    /// and that the vote is signed.
+    pub fn slash(&mut self, slot_hashes: &[SlotHash], slashable_slot: Slot) {
+        //verify that the slot is NOT an ancestor
+        for s in slot_hashes {
+            if s.0 == slashable_sot {
+                //slot is an ancestor
+                return;
+            }
+        }
+        let mut slashable_votes = vec![slashable_slot];
+        let mut test = VoteState::default();
+        slashable_votes.extend(&self.votes());
+        slashable_votes.sort();
+        slashable_votes.dedup();
+        for s in slashable_votes {
+            test.process_vote_unchecked(s);
+        }
+        let not_slashed = if test.votes.last().unwrap() == slashable_slot {
+            //slashable slot is put on top of the stack
+            //the previous vote must be popped
+            self.votes.is_empty()
+                || test.votes[..(test.votes.len() - 1)] == self.votes[..self.votes.len() - 1]
+        } else {
+            //all slashable slots got popped
+            test.votes == self.votes
+        };
+        self.slashed = !not_slashed;
```

Comment:
to store slashed with the epoch for the slot, you need sysvar::epoch_schedule

Reply:
@rob-solana I was thinking once slashed they have to unstake and restake.  the VoteAccount is dead.

Reply:
ok, that's simple.  dead for an epoch.  all unclaimed rewards are gone.